rfcs/text/0008_service_status.md

```diff
@@ -0,0 +1,280 @@
+- Start Date: 2020-03-07
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+A set API for describing the current status of a system (Core service or plugin)
+in Kibana.
+
+# Basic example
+
+```ts
+// Override default behavior and only elevate severity when elasticsearch is not available
+core.status.set(
+  core.status.core$.pipe(core => core.elasticsearch);
+)
+```
+
+# Motivation
+
+Kibana should do as much possible to help users keep their installation in a working state. This includes providing as much detail about components that are not working as well as ensuring that failures in one part of the application do not block using other portions of the application.
+
+In order to provide the user with as much detail as possible about any systems that are not working correctly, the status mechanism should provide excellent defaults in terms of expressing relationships between services and presenting detailed information to the user.
+
+# Detailed design
+
+## Types
+
+```ts
+/**
+ * The current status of a service at a point in time.
+ */
+interface ServiceStatus {
+  /** The current availability level of the service. */
+  level: ServiceStatusLevel;
+  /** A high-level summary of the service status. */
+  summary?: string;
+  /** A more detailed description of the service status. */
+  detail?: string;
+  /** A URL to open in a new tab about how to resolve or troubleshoot the problem */
+  documentationUrl?: string;
+  /** Any JSON-serializable data to be included in the HTTP API response. */
+  meta?: object;
+}
+
+/**
+ * The current "level" of availability of a service.
+ */
+enum ServiceStatusLevel {
+  available,   // everything is working!
+  degraded,    // some features may not be working
```

Comment:
I worry "degraded" has negative connotations that may not be warranted in all cases.

Consider `spaces`:
https://github.com/elastic/kibana/blob/974a51d1042c3ffba0322bd40351aa36971139d8/x-pack/plugins/spaces/kibana.json#L1-L10

If `security` is disabled, the `spaces` feature is still fully-functional from an end-users perspective. They just don't get the additional benefit of access control.

We could have the `spaces` plugin handle this case, and report `available` even if `security` is disabled, but the problem becomes more pervasive if we consider a dependency like `usageCollection`. Now, any plugin which depends on `usageCollection` would need to either:
1) Report a degraded state by default when `usageCollection` is unavailable.
2) Have special-case code to report `available`, even if `usageCollection` is unavailable.

Option 1 feels rather heavy-handed. A Kibana admin should be able to disable usage collection without seeing all of our plugins enter a "degraded" state because they can't phone home anymore.

Option 2 negates the benefit of "Status inheritance", because plugins will need to be selective about when they update.


I started this comment wanting to rename `degraded` to something else, but now I'm wondering if plugins should instead get to choose which dependencies to monitor when inheriting their status. What do you all think?

Reply:
One thing that should improved in clarity here: `unavailable` !== plugin disabled. If an optional dependency is actually disabled, then it does not have a status and therefore it cannot affect the status of plugins that depend on it.

The usageCollection one is a good point, but I think even in that case the behavior is at the very least acceptable. That plugin only depends on Core (for a SavedObjectsRepository) and I can't imagine a reason it would need to go into a degraded or unavailable state except when SavedObjects are having a problem. If SavedObjects are having a problem, then most likely other things _are_ degraded. That said, usageCollection could implement a custom status to never be degraded so that it cannot affect other plugins.

Reply:
I'll also add that "degraded" does not trigger anything in the default Kibana UI (incl. no redirect to the status page). This will only be visible to users who explicitly access this page on purpose.

Since Pulse is becoming a user-facing feature, not only a telemetry mechanism, usageCollection being broken may actually be useful feedback to users.

Reply:
> If an optional dependency is actually disabled, then it does not have a status and therefore it cannot affect the status of plugins that depend on it.

This may be a good idea to have is explicitly state in the RFC?

Reply:
Yep, I have added more clarification to the spec.

Reply:
@legrego did this fully address your concerns?

Reply:
Sorry I somehow missed your replies (thanks for the ping). Yes that satisfies my concerns, thanks!